applications/StructuralMechanicsApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,659 @@
+                for(IndexType i_node_loop = 0; i_node_loop < num_nodes; ++i_node_loop) { // FIXME: Avoid this double loop, extreamily expensive
+                    auto it_node_loop = nodes_array.begin() + i_node_loop;
+                    const SizeType size_elem_neigh = it_node_loop->GetValue(NEIGHBOUR_ELEMENTS).size();
+                    if (it_node_loop->Id() == it_neighbour_nodes->Id() && size_elem_neigh > TDim){
+                        CalculatePatch(it_node, it_node_loop, neighbour_size, sigma_recovered_i);
+                        ++count_i;
+                    }
+                }
+
+                // Average solution from different patches
+                if(count_i != 0)
+                    sigma_recovered = sigma_recovered*(count_i-1)/count_i + sigma_recovered_i/count_i;
+            }
+
+            it_node->SetValue(RECOVERED_STRESS,sigma_recovered);
+            KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 2) << "Recovered sigma: " << sigma_recovered << std::endl;
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculateErrorEstimationAndElementSize(
+    double& rEnergyNormOverall,
+    double& rErrorOverall
+    )
+{
+    // Some initializations
+    const double tolerance = std::numeric_limits<double>::epsilon();
+
+    // Loop over all elements:
+    ElementsArrayType& elements_array = mThisModelPart.Elements();
+    const int num_elem = static_cast<int>(elements_array.size());
+
+    // Compute the error estimate per element
+    double error_overall= 0.0;
+    double energy_norm_overall = 0.0;
+    #pragma omp parallel for reduction(+:error_overall, energy_norm_overall)
+    for(int i_elem = 0; i_elem < num_elem; ++i_elem){
+        auto it_elem = elements_array.begin() + i_elem;
+
+        std::vector<double> error_integration_point;
+        const auto& process_info = mThisModelPart.GetProcessInfo();
+        it_elem->GetValueOnIntegrationPoints(ERROR_INTEGRATION_POINT, error_integration_point, process_info);
+
+        KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 2) << "Error GP:" << error_integration_point << std::endl;
+        
+        // We compute the error overall 
+        double error_energy_norm = 0.0;
+        for(IndexType i = 0;i < error_integration_point.size();++i)
+            error_energy_norm += error_integration_point[i];
+        error_overall += error_energy_norm;
+        error_energy_norm = std::sqrt(error_energy_norm);
+        it_elem->SetValue(ELEMENT_ERROR, error_energy_norm);
+
+
+        // We compute now the energy norm
+        std::vector<double> strain_energy;
+        it_elem->GetValueOnIntegrationPoints(STRAIN_ENERGY, strain_energy, process_info);
+
+        double energy_norm = 0.0;
+        for(IndexType i = 0;i < strain_energy.size(); ++i)
```

Comment:
This and the above for loop could be shortened by replacing with std::accumulate or factoring into a separate function. Putting the `2 * ` out of the loop will also eliminate some flops.